CRAMv3.tex

```diff
@@ -964,23 +964,104 @@ \subsection{\textbf{Core data block}}
+
+\tikzstyle{texto} = [above, text width=8em, text centered]
+ 
+\tikzstyle{encodingArrow} = [thick, decoration={markings, mark=at position
+   1 with {\arrow[semithick]{open triangle 60}}},
+   double distance=3.0pt, shorten >= 5.0pt,
+   preaction = {decorate},
+   postaction = {draw, line width=1.4pt, white, shorten >= 5.0pt}]
+\tikzstyle{innerWhite} = [semithick, white, line width=1.4pt, shorten >= 4.5pt]
+
+\newcommand{\cramRecord}[6]{%
+\begin{pgfonlayer}{background}
+    \path (#1.west |- #1.north)+(-0.5,0.5) node (a1) {};
+    \path (#1.east |- #6.south)+(+0.5,-0.25) node (a2) {};
+    \path[fill=brown!10,rounded corners, draw=black!50, dashed]
+      (a1) rectangle (a2);
+    \path (a1.east |- a1.south)+(1.0,-0.3) node[texto]
+      {\scriptsize\textit{CRAM Record}};
+\end{pgfonlayer}}
+
+\newcommand{\dsbox}[2]{node (p#1) [dsbox]
+   {\scriptsize\textit{#2}}}
+\newcommand{\encodedblock}[2]{node (p#1) [encodedblock]
+   {\scriptsize\textit{#2}}}
+
+\path +(-2.5,-1.5) \dsbox{1}{Bit Flags};
+\path (p1.south)+(0.0,-1.0) \dsbox{2}{Alignment Start};
+\path (p2.south)+(0.0,-1.0) \dsbox{3}{Read Length};
+\path (p3.south)+(0.0,-1.0) \dsbox{4}{Bases};
+\path (p4.south)+(0.0,-1.0) \dsbox{5}{Quality Scores};
+\path (p5.south)+(0.0,-1.0) \dsbox{6}{Tags...};
+
+\cramRecord{p1}{p2}{p3}{p4}{p5}{p6}
+
+\newcommand{\blocks}[4]{%
+\begin{pgfonlayer}{background}
+    \path (#1.west |- #1.north)+(-0.5,0.5) node (a1) {};
+    \path (#1.east |- #1.south)+(+0.5,-0.25) node (a2) {};
+    \path (#2.west |- #2.north)+(-0.5,0.5) node (a3) {};
+    \path (#4.east |- #4.south)+(+0.5,-0.25) node (a4) {};
+    \path[fill=yellow!10, rounded corners, draw=black!50, dashed]
+      (a1) rectangle (a2);
+   \path[fill=yellow!10, rounded corners, draw=black!50, dashed]
+      (a3) rectangle (a4);
+   \path (a1.east |- a1.south)+(0.8,-0.3) node[texto]
+      {\scriptsize\textit{Bit Stream}};
+   \path (a3.east |- a3.south)+(0.8,-0.3) node[texto]
+      {\scriptsize\textit{Byte Streams}};
+\end{pgfonlayer}}
+
+\path (a1.south)+(10.0,-1.0) \encodedblock{7}{Core Data Block};
+\path (p7.south)+(0.0,-2.5) \encodedblock{8}{External Data Block};
+\path (p8.south)+(0.0,-1.0) \encodedblock{9}{External Data Block};
+\path (p9.south)+(0.0,-1.0) \encodedblock{10}{External Data Block...};
+
+\blocks{p7}{p8}{p9}{p10}
+
+\draw[encodingArrow] (p1) + (1.7, -0.0) to (p7);
+\draw[encodingArrow] (p3) + (1.7, 0.0) to (p8);
+\draw[encodingArrow] (p4) + (1.7, 0.0) to (p9);
+\draw[encodingArrow] (p5) + (1.7, 0.0) to (p10);
+
+\filldraw[draw=black, fill=white!10, drop shadow] (1.5, -1.5) ellipse (2.0 and 0.5);
+\filldraw[draw=black, fill=white!10, drop shadow] (1.5, -4.55) ellipse (1.8 and 0.5);
+\filldraw[draw=black, fill=white!10, drop shadow] (1.5, -6.1) ellipse (1.8 and 0.5);
+\filldraw[draw=black, fill=white!10, drop shadow] (1.5, -7.6) ellipse (1.8 and 0.5);
+
+\node [right=of p1, scale=0.7, text centered, text width=14em] {Core (Bit) Encodings (Huffman, RANS, etc.)};
```

Comment:
RANS isn't a bit encoding; it only applies to external blocks.  Instead "Huffman, Beta, etc." would be appropriate.

I'm also a little wary of the layout as it sort of implies specifics which the old picture didn't.  We previously had big fat arrows that spanned multiple items on the left and weren't well aligned with them.  The new one looks to have a deliberate one to one alignment with specific types of data.  While this is just an example, I'd be cautious not to hint that specific types are encoding in specific ways.

If  we wish to keep this image, we'll want to change the legend from things like "how a CRAM record _is_ distributed" to "how a CRAM record _may be_ distributed".

PS. More complex Tikz work than I've done so far - a good opener!   One of the reasons I didn't do this was because it'd mean spending more time, which I didn't have, learning Tikz. :-)

Reply:
The flip side to the above argument, if we are happy with a diagram showing an example rather than describing the basic concepts without implying anything concrete,  then we should perhaps be more bold and give a realistic example.

Eg using specific types of byte encodings (EXTERNAL,  BYTE_ARRAY_STOP, etc) on obvious numerics and strings.  We could then also add appropriate examples of compression codecs.  Eg BYTE_ARRAY_STOP on read identifiers to an external block compressed using bzip2,  EXTERNAL for quality scores compressed with rANS, and so on.  Maybe pick Bit flags, alignment start for HUFFMAN and BETA encodings.

We need to keep a clear separation of encoding (which may be byte based or bit based) and compression codec (as-is, gzip, bzip2, lzma, rans).  Especially as this is likely something we'll expand on in future.

Reply:
Ok, thx for the comments. Point taken on RANS, thats in the wrong place. Also, I had tried to make the left side of the arrows start from the `CRAM Record` box rather than the individual fields, and the right side end at the individual blocks, but perhaps thats too subtle. I'll do another pass.